### ISSUE TITLE
Override existing checks with a plugin 

### ISSUE THREAD

**Issue by kasium** [PyCQA/pylint#5274]
### Current problem

Hey guys,

I use [pyinvoke](https://www.pyinvoke.org/) a lot, but pylint doesn't like some stuff a lot.
In invoke so called tasks are declared like this

```python
@task
def my_task(context):
  print("Hello World")
```

Please note the `context` argument: this argument is mandatory for each task and must be named like this. Pylint detects completely a right an `unused-argument`.

For my project I wrote a small pylint plugin, which finds all functions decorated with `task` and an `context` argument and ignores the `unused-argument`. However, this is done by patching `VariablesChecker.add_message`. This doesn't look nice to me:

```python
def register(linter):
    original = VariablesChecker.add_message

    def _new_add_message(
        self, msgid, line = None, node = None, args= None, confidence= None, col_offset= None,
    ):
        if (
            msgid == "unused-argument"
            and isinstance(node, AssignName)
            and node.name == "context"
            and isinstance(node.parent, Arguments)
            and isinstance(node.parent.parent, FunctionDef)
            and "invoke.tasks.task" in node.parent.parent.decoratornames()
        ):
            return
        original(self, msgid, line, node, args, confidence, col_offset)

    VariablesChecker.add_message = _new_add_message
```

### Desired solution

A way to improve certain existing checks, e.g. by override

### Additional context

[Pylint-pytest](https://pypi.org/project/pylint-pytest/) uses a similar way as I do

**Comment by Pierre-Sassoulas** [comment]
Why isn't it possible to inherit from ``VariablesChecker`` and override the ``add_message`` function ?

**Comment by kasium** [comment]
Thanks for the fast answer. And how do I then tell pylint to use the new `VariablesChecker` instead of the old one? I don't want to add new checks (as described [here](https://pylint.pycqa.org/en/latest/how_tos/custom_checkers.html)) but improving an existing one

**Comment by Pierre-Sassoulas** [comment]
OK I understand the problem now. Maybe we could create a ``register_inherited_checker`` function ?

```python
class PyinvokeVariablesChecker(VariablesChecker): 

    def add_message(
        self,
        msgid: str,
        line: Optional[int] = None,
        node: Optional[nodes.NodeNG] = None,
        args: Any = None,
        confidence: Optional[Confidence] = None,
        col_offset: Optional[int] = None,
    ) -> None:
        if (
            msgid == "unused-argument"
            and isinstance(node, AssignName)
            and node.name == "context"
            and isinstance(node.parent, Arguments)
            and isinstance(node.parent.parent, FunctionDef)
            and "invoke.tasks.task" in node.parent.parent.decoratornames()
        ):
            return
        return super(...)

register_inherited_checker(checker=PyinvokeVariablesChecker, parent=VariablesChecker)
```
I probably need to check if something like this exists first but if pylint-pytest does it this way it probably does not exists.

**Comment by kasium** [comment]
great! thanks a lot

**Comment by Pierre-Sassoulas** [comment]
Closing as answered.

**Comment by kasium** [comment]
Is register_inherited_checker something pylint can provide in the future?

**Comment by Pierre-Sassoulas** [comment]
Hmm thank you for your reaction I was too trigger happy with this one, it's an enhancement not a question.

**Comment by kasium** [comment]
Not sure about the process, but here are some ideas:
Even if register_inherited_checker is a nice idea, I guess this can lead to issues. If two plugins want to overwrite the same parent, it can lead to confusion. Which plugin comes first and when is which message emitted?

To avoid this, maybe the extension can happen on message basis. A plugin can register a custom handler for a given message:
```py
register_message_handler("unused_argument", my_handler)
```

If then "unused_argument" is emitted and a handler is registered, the handler receives the current node. It can then decide to do additional conditions (as in my initial example) or e.g. add additional stuff. At the end, the handler can return a boolean value. This value indicates, that the message should be emitted by pylint or not. If yes (return True), then we need to call `self.add_message`. This should just avoid, that plugins need to have an else case where they call `self.add_message` all the time. If False is returned, no further processing is done.
If more than one handler is installed, all handlers are called independently of the return values.

The plugin code
```py
def handler(node):
  if node.name == "context":
    # just ignore the message
  return False
return True
register_message_handler("unused_argument", handler)
```

The plyint processing:
```py
# dict of message id linking to a list if handlers
handlers: dict[str, list[Callable]] = {}

def _add_message(msgid, node):
  should_add = True  
  for handler in handler.get(msgid, []):
    should_add &= handler(node)
  if should_add:
  self.add_message(...)
```

What do you think?

**Comment by DanielNoord** [comment]
I have been thinking about this as well. I think we should be very careful with the definition of the handle callable. For example I have seen the need for `linter` as well so I defined my handler as `Callable[[PyLinter | None, NodeNG | None], None]`.

I think we should try and see if there are other corner cases and perhaps relevant information that should also be passed.

**Comment by kasium** [comment]
Jep, they do this by entrypoint, quite fancy

**Comment by kasium** [comment]
Hi all,

maybe to continue the discussion again: Entry-point are great and flake8 works perfect with it. However, pylint is more complicated and the current "plugin api" doesn't use entry points, so I'm not sure if this would be the right time to start this.
I think that this message handler API would be a good way forward. It's flexible, extensible and somehow easy to implement. 

Regarding the callable API: I would recommend going with either a single object or forced kwargs.
```py
# single object
message = Message(id=..., node=...)

def my_message_handler(message: Message): ...

# kwargs; can be enforced by using a Protocol
def my_message_handler(*, id:str, node, **kwargs: Any)
```

Both solutions would allow pylint to pass more and more data into the function, deprecate attributes, etc. What do you think?

**Comment by Pierre-Sassoulas** [comment]
You're right that implementing the entrypoints powered plugins like pytest and flake8 is another issue that would need to be discussed elsewhere.

Regarding the APi you propose I'd rather have a single object than a kwargs. It's hard to know what's supposed to be in a kwargs and it would make the API unclear. The Message object already exists I think, but it might not be the object we need.

**Comment by DanielNoord** [comment]
For the pytest plug-in I worked in I just overwrote the `add_message` method. You really want access to the node so and all other context so I don't think you can avoid that.

I do agree we need a better system for this.

**Comment by kasium** [comment]
@DanielNoord can you please explain why you think the overwrite is still needed? I did the same with https://github.com/SAP/invoke-plugin-for-pylint and I guess that the approach in https://github.com/pylint-dev/pylint/issues/5274#issuecomment-1537103588 would work here

**Comment by DanielNoord** [comment]
Oh your suggestion could also work, but like Pierre said we already have a `Message` object. I'm not sure how feasible it would be to add `node` to it (not all `Messages` have a `node` for example), so perhaps it is better to not mangle the two and just expose a separate interface?
